### ISSUE TITLE
exp/horizonclient: returned Horizon errors have problematic types

### ISSUE THREAD

**Issue by ire-and-curses** [stellar/go#1100]
When `horizonclient` generates a protocol level error, attempting to access the contents fails with a type assertion error.

Looking at the contents of `ResultCodes` in `error.go`, it looks like the logic is wrong.

 First, here's a modified version of that function that successfully parses the data:

```
// ResultCodes extracts a result code summary from the error, if possible.
func (herr *Error) ResultCodes() (*hProtocol.TransactionResultCodes, error) {

    raw, ok := herr.Problem.Extras["result_codes"]
    if !ok {
        return nil, ErrResultCodesNotPopulated
    }

    // rawB, ok := raw.([]byte)
    rawB := raw.(map[string]interface{})

    decoded := make(map[string]string)
    for k, v := range rawB {
        strKey := fmt.Sprintf("%v", k)
        strValue := fmt.Sprintf("%v", v)
        fmt.Println(reflect.TypeOf(strKey), reflect.TypeOf(strValue))
        fmt.Printf("%v -> %v\n", strKey, strValue)

        decoded[strKey] = strValue
    }
    fmt.Println(raw)
    fmt.Printf("%v\n", rawB)

    return nil, nil
```

Output:

```
string string
transaction -> tx_failed
string string
operations -> [op_already_exists]
map[transaction:tx_failed operations:[op_already_exists]]
map[transaction:tx_failed operations:[op_already_exists]]
```

So this is telling us a few things:
1) The error has already been decoded, and stored in a map
2) The map is really of type `map[string]string`
3) The map is being passed in as an empty interface which must be cast
4) The values of the map are also empty interfaces which must be cast
Tracing down through the `submit` function, I see that the decoding happens in `decodeResponse` in `internal.go`.

The thing that is confusing me is that we don't really want a `map` at all. We know what Horizon will return, and we want the defined type `hProtocol.TransactionResultCodes`. So I'm not sure how we got to this point here.

One other thing is that the definition of `hProtocol.TransactionResultCodes` is

```
type TransactionResultCodes struct {
    TransactionCode string   `json:"transaction"`
    OperationCodes  []string `json:"operations,omitempty"`
}
```

So OperationCodes should be a slice of strings. But it seems to be just a string instead. I think this may be a parsing error in the initial decoding.

So issues are
1) The code is attempting to decode the JSON twice
2) The decoding is going into a very generic map, but a struct would be preferable
3) The map value for `operations` seems to be a `string`, but should be `[]string`

The same is probably true for the other fields (e.g. `ResultString`, `Envelope`).

**Comment by ire-and-curses** [comment]
This is blocking #1070.

**Comment by poliha** [comment]
Hey, thanks for looking into this. The issues you have pointed out are correct.  
The main issue is that we were not casting to the right type, and were still trying to decode a json response that was already decoded.  
It was okay in the old client because it was `map[string]json.RawMessage` so we could decode `json.RawMessage`.  It's failing here because its now `map[string]interface{}`
I have worked on this and should be fixed in #1104

**Comment by poliha** [comment]
Resolved in #1104
